packages/headless-react/src/hooks/controller-hook.ts

```diff
@@ -0,0 +1,159 @@
+    if (typeof value === 'function') {
+      (methods as Record<string, unknown>)[key] = value.bind(obj);
+    } else if (typeof value === 'object') {
+      bindMethods(value, (obj as Record<string, object>)[key]);
+    }
+  });
+  return methods;
+}
+
+function getOnStateChanged(subscribe: (listener: () => void) => Unsubscribe) {
+  return (listener: () => void) => {
+    let firstCall = true;
+    return subscribe(() => {
+      if (firstCall) {
+        firstCall = false;
+        return;
+      }
+      listener();
+    });
+  };
+}
+
+function buildController(
+  builder: (engine: CoreEngine, props?: object) => Controller,
+  engine: CoreEngine,
+  props?: object
+) {
+  const headlessController = builder(engine, props);
+  const {subscribe: _, state: __, ...methods} = headlessController;
+  return {
+    controller: bindMethods(methods, headlessController),
+    getStateSnapshot: () => structuredClone(headlessController.state),
+    onStateChanged: getOnStateChanged((listener) =>
+      headlessController.subscribe(listener)
+    ),
+  };
+}
+
+/**
+ * Same as [useSyncExternalStore](https://reactjs.org/docs/hooks-reference.html#usesyncexternalstore), but doesn't infinitely re-render with a memoized store.
+ */
+function useSyncMemoizedStore<T>(
+  subscribe: (listener: () => void) => () => void,
+  getSnapshot: () => T
+) {
+  const state = useRef<T | null>(null);
+  const [, render] = useReducer((renderCount: number) => renderCount + 1, 0);
+  useEffect(
+    () =>
+      subscribe(() => {
+        state.current = getSnapshot();
+        render();
+      }),
+    [subscribe, getSnapshot]
+  );
+  if (useHasChanged([subscribe, getSnapshot]) || state.current === null) {
+    state.current = getSnapshot();
+  }
+  return state.current as T;
+}
+
+export function useController<
+  TController extends Controller,
+  TEngine extends CoreEngine
+>(
+  builder: (engine: TEngine) => TController,
+  engine: TEngine
+): UseControllerReturnType<TController>;
+export function useController<
+  TController extends Controller,
+  TEngine extends CoreEngine,
+  TProps extends object
+>(
+  builder: (engine: TEngine, props: TProps) => TController,
+  engine: TEngine,
+  props: TProps
+): UseControllerReturnType<TController>;
+export function useController(
+  builder: (engine: CoreEngine, props?: object) => Controller,
+  engine: CoreEngine,
+  props?: object
+): UseControllerReturnType<Controller> {
+  if (useHasChanged([builder])) {
+    console.error(
+      'A Coveo Headless controller builder was changed in between renders. This can happen when a custom callback is passed as a builder instead of an build function imported directly from Headless.'
+    );
+  }
+  const {controller, getStateSnapshot, onStateChanged} = useDeepCompareMemo(
+    () => buildController(builder, engine, props),
+    props ?? {},
+    [builder, engine]
+  );
+  const state = useSyncMemoizedStore(onStateChanged, getStateSnapshot);
```

Comment:
React made the [`useSyncExternalStore`](https://reactjs.org/docs/hooks-reference.html#usesyncexternalstore) hook, which does pretty much exactly what we need:
* Has an initial state that updates when the store is changed (in this case, `onStateChanged`/`getStateSnapshot` are the store).
* Re-renders the component with the new state whenever the listener is called.

I even tested `useSyncExternalStore` with those two exact parameters.

However, `useSyncExternalStore` causes an infinite re-rendering loop, like it's detecting `onStateChanged` and/or `getStateSnapshot` to be different on every single render (even when other hooks don't detect a change for the same dependencies). `useSyncExternalStore` is meant to subscribe to a store that was statically imported from another module, not a memoized value, so I believe that's why I'm not finding much about the issue online. I tried messing with `useSyncExternalStore` for over an hour.

For now, this will do. I'll add a lot of tests, and we can update this in the future.